File: android/app/src/main/java/mx/medsi/medsicheck/RNTEventBridge.kt
/*
 *              Copyright (c) 2016-2023, Nuralogix Corp.
 *                      All Rights reserved
 *
 *      THIS SOFTWARE IS LICENSED BY AND IS THE CONFIDENTIAL AND
 *      PROPRIETARY PROPERTY OF NURALOGIX CORP. IT IS
 *      PROTECTED UNDER THE COPYRIGHT LAWS OF THE USA, CANADA
 *      AND OTHER FOREIGN COUNTRIES. THIS SOFTWARE OR ANY
 *      PART THEREOF, SHALL NOT, WITHOUT THE PRIOR WRITTEN CONSENT
 *      OF NURALOGIX CORP, BE USED, COPIED, DISCLOSED,
 *      DECOMPILED, DISASSEMBLED, MODIFIED OR OTHERWISE TRANSFERRED
 *      EXCEPT IN ACCORDANCE WITH THE TERMS AND CONDITIONS OF A
 *      NURALOGIX CORP SOFTWARE LICENSE AGREEMENT.
 */

package mx.medsi.medsicheck

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import mx.medsi.medsicheck.ui.MainActivity
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import org.json.JSONObject
import java.util.Locale

enum class Action {
    StartMeasurement,
    SynchronizeAppConfiguration,
    SynchronizeConfiguration,
    SynchronizeUIConfiguration,
    StopCommonObserving,
    StopResultsObserving;

    val stringValue = this.toString().replaceFirstChar { it.lowercase(Locale.getDefault()) }
}

enum class Event {
    AnuraMeasurementPageDidLoad,
    AnuraMeasurementPageDidAppear,
    AnuraMeasurementPageDidDisappear,
    AnuraMeasurementPageIsReadyToMeasure,
    AnuraMeasurementPageDidStartMeasuring,
    AnuraMeasurementPageDidFinishMeasuring,
    AnuraMeasurementGetResultsSuccess,
    AnuraMeasurementGetResultsFailure;

    val stringValue = this.toString().replaceFirstChar { it.lowercase(Locale.getDefault()) }
}

class RNTEventBridge(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    companion object {
        lateinit var reactContext: ReactApplicationContext

        @JvmStatic fun sendCommomEvent(name: String, params: WritableMap?) {
            val newParams = Arguments.createMap().apply {
                putString("nativeActionName", name)
                putMap("data", params)
            }
            sendEvent(reactContext, "eventReminder_common", newParams)
        }

        @JvmStatic fun sendResultsEvent(name: String, params: WritableMap?) {
            val newParams = Arguments.createMap().apply {
                putString("nativeActionName", name)
                putMap("data", params)
            }
            sendEvent(reactContext, "eventReminder_results", newParams)
        }

        private fun sendEvent(reactContext: ReactContext, eventName: String, params: WritableMap?) {
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(eventName, params)
        }
    }

    override fun getName() = "RNTEventBridge"

    @ReactMethod
    fun doSomething(eventName: String, property: String?) {
        when (eventName) {
            Action.StartMeasurement.stringValue -> {
                (MainActivity.currentContext as MainActivity).startMeasurementActivity(jsonToMap(property))
            }
            Action.SynchronizeAppConfiguration.stringValue -> {
                AppConfig.synchronizeAppConfiguration(jsonToMap(property))
            }
            Action.SynchronizeConfiguration.stringValue -> {
                MeasurementConfig.synchronizeMeasurementConfiguration(jsonToMap(property))
            }
            Action.SynchronizeUIConfiguration.stringValue -> {
                MeasurementUIConfig.synchronizeMeasurementUIConfiguration(jsonToMap(property))
            }
        }
    }

    private fun jsonToMap(jsonString: String?): Map<String, Any>? {
        if (jsonString == null) return null
        val json = JSONObject(jsonString)
        val map = mutableMapOf<String, Any>()
        for (key in json.keys()) {
            val value = json.get(key)
            map[key] = value
        }
        return map
    }

    @ReactMethod
    fun anura_startCommonObserving() {}

    @ReactMethod
    fun anura_startResultsObserving() {}

    @ReactMethod
    fun addListener(eventName: String) {}

    @ReactMethod
    fun removeListeners(count: Int) {}
}

File: ios/Main/RNTEventBridge.m
//
//  Copyright (c) 2016-2023, Nuralogix Corp.
//  All Rights reserved
//  THIS SOFTWARE IS LICENSED BY AND IS THE CONFIDENTIAL AND
//  PROPRIETARY PROPERTY OF NURALOGIX CORP. IT IS
//  PROTECTED UNDER THE COPYRIGHT LAWS OF THE USA, CANADA
//  AND OTHER FOREIGN COUNTRIES. THIS SOFTWARE OR ANY
//  PART THEREOF, SHALL NOT, WITHOUT THE PRIOR WRITTEN CONSENT
//  OF NURALOGIX CORP, BE USED, COPIED, DISCLOSED,
//  DECOMPILED, DISASSEMBLED, MODIFIED OR OTHERWISE TRANSFERRED
//  EXCEPT IN ACCORDANCE WITH THE TERMS AND CONDlITIONS OF A
//  NURALOGIX CORP SOFTWARE LICENSE AGREEMENT.
//

#import "RNTEventBridge.h"
#import "MedsiCheckDev-Swift.h" 
//#import "MedsiCheck-Swift.h"
// #import "MedsiCheckStaging-Swift.h"

NSString *const EventReminderCommon = @"eventReminder_common";
NSString *const EventReminderResults = @"eventReminder_results";
NSString *const NativeActionName = @"nativeActionName";

typedef NS_ENUM(NSUInteger, Action) {
  StartMeasurement,
  SynchronizeAppConfiguration,
  SynchronizeConfiguration,
  SynchronizeUIConfiguration,
  StopCommonObserving,
  StopResultsObserving,
};

NSString * _Nonnull ActionName[] = {
  [StartMeasurement] = @"startMeasurement",
  [SynchronizeAppConfiguration] = @"synchronizeAppConfiguration",
  [SynchronizeConfiguration] = @"synchronizeConfiguration",
  [SynchronizeUIConfiguration] = @"synchronizeUIConfiguration",
  [StopCommonObserving] = @"stopCommonObserving",
  [StopResultsObserving] = @"stopResultsObserving",
};

@implementation RNTEventBridge

RCT_EXPORT_MODULE()

- (NSArray<NSString *> *)supportedEvents
{
  return @[EventReminderCommon, EventReminderResults];
}

RCT_EXPORT_METHOD(anura_startCommonObserving)
{
  [self startObserving:EventReminderCommon body:nil];
}

RCT_EXPORT_METHOD(anura_startResultsObserving)
{
  [self startObserving:EventReminderResults body:nil];
}


RCT_EXPORT_METHOD(doSomething:(NSString *)eventName property:(NSString *)property)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    if ([eventName isEqual:ActionName[StartMeasurement]]) {
      [[StartSession shared] startWithProperties:property];
    } else if ([eventName isEqual:ActionName[SynchronizeConfiguration]]) {
      [[StartSession shared] synchronizeConfiguration:property];
    } else if ([eventName isEqual:ActionName[SynchronizeUIConfiguration]]) {
      [[StartSession shared] synchronizeUIConfiguration:property];
    } else if ([eventName isEqual:ActionName[SynchronizeAppConfiguration]]) {
      [AppConfig synchronizeAppConfiguration:property];
    }
  });
}

- (void)startObserving:(NSString *)name body:(id)object {
  NSLog(@"rn--iOS...startObserving(%@)", name);
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(eventSend:) name:name object:nil];
}

- (void)stopObserving:(NSString *)name body:(id)object {
  NSLog(@"rn--iOS...stopObserving(%@)", name);
  [[NSNotificationCenter defaultCenter] removeObserver:self name:name object:nil];
}

- (void)stopAllObserving {
  NSLog(@"rn--iOS...stopAllObserving");
  [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)eventSend:(NSNotification *)noti
{
  dispatch_async(dispatch_get_main_queue(), ^{
    if ([noti.object isKindOfClass:[NSDictionary class]] || noti.object == nil) {
      NSDictionary *dic = noti.object;
      NSString *actionName = dic[NativeActionName];
      if ([actionName isEqual:ActionName[StopCommonObserving]] || [actionName isEqual:ActionName[StopResultsObserving]]) {
        [self stopObserving:noti.name body:noti.object];
      }
      [self sendEventWithName:noti.name body:dic];
    }
  });
}

+ (void)sendCommonEvent:(NSString *)name body:(nullable id)object
{
  NSMutableDictionary *dic = [NSMutableDictionary dictionary];
  if ([object isKindOfClass:[NSDictionary class]] || [object isKindOfClass:[NSArray class]]) {
    [dic setValue:object forKey:@"data"];
    [dic setValue:name forKey:NativeActionName];
  } else if (object == nil) {
    [dic setValue:name forKey:NativeActionName];
  }
  
  [[NSNotificationCenter defaultCenter] postNotificationName:EventReminderCommon object:dic];
}

+ (void)sendResultsEvent:(NSString *)name body:(nullable id)object
{
  NSMutableDictionary *dic = [NSMutableDictionary dictionary];
  if ([object isKindOfClass:[NSDictionary class]] || [object isKindOfClass:[NSArray class]]) {
    [dic setValue:object forKey:@"data"];
    [dic setValue:name forKey:NativeActionName];
  } else if (object == nil) {
    [dic setValue:name forKey:NativeActionName];
  }
  
  [[NSNotificationCenter defaultCenter] postNotificationName:EventReminderResults object:dic];
}

+ (void)stopCommonObserving {
  [[NSNotificationCenter defaultCenter] postNotificationName:EventReminderCommon object:@{NativeActionName: ActionName[StopCommonObserving]}];
}

+ (void)stopResultsObserving {
  [[NSNotificationCenter defaultCenter] postNotificationName:EventReminderResults object:@{NativeActionName: ActionName[StopResultsObserving]}];
}
@end
